engine/runtime/src/main/java/org/enso/interpreter/runtime/error/WithWarnings.java

```diff
@@ -158,22 +193,42 @@ public int hashCode(Object o) {
   }
 
   @CompilerDirectives.TruffleBoundary
-  private EconomicSet<Warning> createSetFromArray(Warning[] entries) {
+  private EconomicSet<Warning> createSetFromArray(int maxWarnings, Warning[] entries) {
     EconomicSet<Warning> set = EconomicSet.create(new WarningEquivalence());
-    set.addAll(Arrays.stream(entries).iterator());
+    for (int i=0; i<entries.length; i++) {
+      if (set.size() == maxWarnings) {
+        return set;
+      }
+      set.add(entries[i]);
+    }
     return set;
   }
 
+  private EconomicSet<Warning> cloneSetAndAppend(int maxWarnings, EconomicSet<Warning> initial, Warning[] entries) {
+    return initial.size() == maxWarnings ? initial : cloneSetAndAppendSlow(maxWarnings, initial, entries);
+  }
+
   @CompilerDirectives.TruffleBoundary
-  private EconomicSet<Warning> cloneSetAndAppend(EconomicSet<Warning> initial, Warning[] entries) {
+  private EconomicSet<Warning> cloneSetAndAppendSlow(int maxWarnings, EconomicSet<Warning> initial, Warning[] entries) {
```

Comment:
Isn't this exactly the right place to check if we reached max warnings count? We could report that and rely on this value.

IMO it should be computed in one place, scattering it all over the codebase will result in some place at some point being wrong and we will be investigating a bug "why sometimes the limit message comes up and sometimes not?" and I'd really like to avoid it.